### DISCUSSION TITLE
Direct Forwarder exception: 'Received chunk header length could not be parsed'

### DISCUSSION THREAD

**Discussion by lahiru119** [dotnet/yarp#1885] [Q&A] [answered]
Hi,

I have developed a simple Direct Forwarding solution, using the basic example. 

```c#
endpoints.Map("/{**catch-all}", async httpContext =>
                {
                    var error = await forwarder.SendAsync(httpContext, "https://example.com", httpClient, requestOptions, transformer);                    
                    if (error != ForwarderError.None)
                    {
                        var errorFeature = httpContext.Features.Get<IForwarderErrorFeature>();
                        var exception = errorFeature.Exception;
                    }
                });
```

I am using this to forward the requests to different destinations, depend on some conditions. For most sites it works great, but for one I get the following exception. It would be great if anyone has any suggestions on how to resolve this. 

_{"Received chunk header length could not be parsed: '3C-21-44-4F-43-54-59-50-45-20-48-54-4D-4C-20-50-55-42-4C-49-43-20-22-2D-2F-2F-57-33-43-2F-2F-44-54-44-20-48-54-4D-4C-20-34-2E-30-31-20-54-72-61-6E-73-69-74-69-6F-6E-61-6C-2F-2F-45-4E-22-3E'."}_

Thank you.

**Comment by Tratcher** [comment, marked as answer]
Can you capture a wireshark trace? We'd need to see what headers accompany that response. That error implies that there was a `Transfer-Encoding: chunked` header, but that the body wasn't formatted correctly. Those bytes decode as follows, which is not using the chunking format.
`<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">`

**Reply by lahiru119** [reply]
Hi Tratcher,

Thanks for getting back to me. 
Yes, there is a Transfer-Encoding: chunked header. Following is the wireshark capture of this. And that is all in the wireshark capture. But please let me know if you like to see the full capture, or the full html source code of this requesting webpage. 

Thanks again. 

```http
GET /main.html HTTP/1.1
Host: 192.168.100.20
Accept: text/html,application/xhtml+xml,application/xml;q=0.9,image/webp,image/apng,*/*;q=0.8,application/signed-exchange;v=b3;q=0.9
X-Forwarded-For: ::1
User-Agent: Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/106.0.0.0 Safari/537.36 Edg/106.0.1370.37
Accept-Encoding: gzip, deflate, br
Accept-Language: en-GB,en;q=0.9,en-US;q=0.8
Cookie: lang=en
sec-fetch-site: none
sec-fetch-mode: navigate
sec-fetch-user: ?1
sec-fetch-dest: document
sec-ch-ua: "Chromium";v="106", "Microsoft Edge";v="106", "Not;A=Brand";v="99"
sec-ch-ua-mobile: ?0
sec-ch-ua-platform: "Windows"
X-Forwarded-Host: localhost:7232
X-Forwarded-Proto: https
traceparent: 00-97f2c42e4c68bbcd94a9655dcee0b383-de532c8d12e46c09-00

HTTP/1.1 200 OK
Server: Rapid Logic/1.1
MIME-version: 1.0
Date: Thu Jan  8 02:48:26 1970 GMT
Content-Type: text/html; charset=UTF-8
Transfer-Encoding: chunked
Connection: close
Pragma: no-cache
Cache-Control: no-cache
Set-Cookie: MFPSESSIONID=0200C7EFF3B0A014D9EFBD73A725582AA633CBEB11C41712F061202210111257206502
Extend-sharp-setting-status: 0

<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html  lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta http-equiv="Content-Script-Type" content="text/javascript">
<meta http-equiv="Content-Style-Type" content="text/css">
<link rel="stylesheet" type="text/css" href="style1.css" charset="UTF-8">

<title>Top Page - MX-3110N</title>

	<script type="text/javascript" charset="UTF-8">
	<!--
	function loginSubmit(mode) {
		var locationURL;

		if(mode == 0) {
			locationURL = "login.html?" + location.pathname;
			location.href = locationURL;
		}
		else {
			location.href = "login.html?logoffSubmit";
		}
	}
	//-->
	</script>
<!-- javascript //-->
<script type="text/javascript" charset="UTF-8" src="webpage.js"></script>
<script type="text/javascript" charset="UTF-8">
<!--
sflg = 0;
function validate( name ){
	if (sflg == 0){
		sflg = 1;
		document.main.action.value = name;
		ordinate = document.getElementsByName("ordinate");
		if (ordinate.length > 0){
			document.main.ordinate.value = document.body.scrollTop || document.documentElement.scrollTop;
		}
		document.main.submit();
	}
}
//-->
</script>
```

**Reply by MihaZupan** [reply]
The server is replying with invalid HTTP. It is specifying `Transfer-Encoding: chunked`, but isn't actually applying the chunking.
YARP doesn't know what to do with this invalid data and reports an error.

There is no way to control this internal behavior of `HttpClient` (the part that parses the server's response), and we would not expose options to do so in this case.

Do you have the option to change/fix the server behavior here (either add proper chunk framing, or avoid adding the header)?
If you can't influence the server at all, the only option would be to manually rewrite the server response before it reaches YARP (a hacky workaround to handle a misbehaving server).

**Reply by karelz** [reply]
@lahiru119 do you know what is the server "Rapid Logic/1.1". Per @MihaZupan's internet search it may be some kind of embedded device.
Is any other HTTP Client (e.g. browser) working correctly with this server?

There may be similar problem with it from 2006 reported on curl: https://curl.se/mail/lib-2006-04/0046.html

**Reply by lahiru119** [reply]
Hi All,

Thanks for your replies. 

@MihaZupan we are unable change/fix the server side of things. 

@karelz Yes, the server is on an embedded device. And we can successfully load the web page if we access directly via any web browser. 

I will look into your suggestions. 

Thanks again. Really appreciate your help.

**Reply by karelz** [reply]
Given that this is server problem and it is violating RFC, we do not plan to address it in HttpClient.

**Reply by lahiru119** [reply]
Hi @karelz ,

I managed to get it working by following the link you provided (https://curl.se/mail/lib-2006-04/0046.html). Thanks a lot for that. 

And thanks to everyone too.
